src/validation.cpp

```diff
@@ -4196,6 +4202,28 @@ static bool ContextualCheckBlockHeader(const CBlockHeader& block, BlockValidatio
     return true;
 }
 
+/**
+ * We want to enforce certain rules (specifically the 64-byte transaction check)
+ * before we call CheckBlock to check the merkle root. This allows us to enforce
+ * malleability checks which may interact with other CheckBlock checks.
+ * This is currently called both in AcceptBlock prior to writing the block to
+ * disk and in ConnectBlock.
+ * Note that as this is called before merkle-tree checks so must never return a
+ * non-malleable error condition.
+ */
+static bool ContextualBlockPreCheck(const CBlock& block, BlockValidationState& state, const ChainstateManager& chainman, const CBlockIndex* pindexPrev)
+{
```

Comment:
throwing this out there in case it makes sense based on the docs
```Suggestion
{
    Assume(!block.fChecked); 
```
is it important to run this check specifically before the merkle checks?

Reply:
If you want to remove the call to `CheckBlock` prior to `AcceptBlock` in `ProcessNewBlock` then it's important to issue a `BLOCK_MUTATED` error for blocks with all 64-byte txs at a similar time to the merkle checks, as otherwise you'll get a real error which `AcceptBlock` will cache, preventing you from seeing the real block.

If you don't want to remove that call, then it matters a lot less where the 64b tx check is made.

As far as `block.fChecked` goes, seems to me it's more the reverse: you'd want `ContextualBlockPreCheck()` to set `block.m_prechecked = true;` and `CheckBlock()` to verify that's already been set?